pkg/comp-functions/functions/vshnredis/backup.go

```diff
@@ -0,0 +1,374 @@
+package vshnredis
+
+import (
+	"context"
+	_ "embed"
+	"encoding/json"
+
+	xhelm "github.com/crossplane-contrib/provider-helm/apis/release/v1beta1"
+	xkube "github.com/crossplane-contrib/provider-kubernetes/apis/object/v1alpha1"
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	k8upv1 "github.com/k8up-io/k8up/v2/api/v1"
+	"github.com/sethvargo/go-password/password"
+	appcatv1 "github.com/vshn/appcat/apis/v1"
+	vshnv1 "github.com/vshn/appcat/apis/vshn/v1"
+	"github.com/vshn/appcat/pkg/comp-functions/functions/common"
+	"github.com/vshn/appcat/pkg/comp-functions/runtime"
+	"gopkg.in/yaml.v3"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/utils/pointer"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+)
+
+const (
+	credentialSecretName = "backup-bucket-credentials"
+	k8upRepoSecretName   = "k8up-repository-password"
+	k8upRepoSecretKey    = "password"
+	backupScriptCMName   = "backup-script"
+)
+
+type backupResFunc func(ctx context.Context, comp *vshnv1.VSHNRedis, iof *runtime.Runtime) error
+
+type backupCompStep struct {
+	Name    string
+	ResFunc backupResFunc
+}
+
+var backupResources = []backupCompStep{
+	{
+		Name:    "objectbucket",
+		ResFunc: createObjectBucket,
+	},
+	{
+		Name:    "credential observer",
+		ResFunc: createObjectBucketCredentialObserver,
+	},
+	{
+		Name:    "repository password",
+		ResFunc: createRepositoryPassword,
+	},
+	{
+		Name:    "k8up schedule",
+		ResFunc: createK8upSchedule,
+	},
+	{
+		Name:    "backup cm",
+		ResFunc: createScriptCM,
+	},
+	{
+		Name:    "adjust helm values",
+		ResFunc: adjustHelmValues,
+	},
+}
+
+//go:embed script/backup.sh
+var redisBackupScript string
+
+// AddBackup creates an object bucket and a K8up schedule to do the actual backup.
+func AddBackup(ctx context.Context, iof *runtime.Runtime) runtime.Result {
+
+	l := controllerruntime.LoggerFrom(ctx)
+
+	comp := &vshnv1.VSHNRedis{}
+	err := iof.Desired.GetComposite(ctx, comp)
+	if err != nil {
+		return runtime.NewFatalErr(ctx, "failed to parse composite", err)
+	}
+
+	common.SetRandomSchedules(&comp.Spec.Parameters.Backup, &comp.Spec.Parameters.Maintenance)
+
+	err = iof.Desired.SetComposite(ctx, comp)
+	if err != nil {
+		return runtime.NewFatalErr(ctx, "failed to set composite", err)
+	}
+
+	for _, res := range backupResources {
+		l.Info("running backup step", "name", res.Name)
+		err := res.ResFunc(ctx, comp, iof)
+		if err != nil {
+			return runtime.NewFatalErr(ctx, "error during backup step: "+res.Name, err)
+		}
+	}
+
+	err = createScriptCM(ctx, comp, iof)
+	if err != nil {
+		return runtime.NewFatalErr(ctx, "cannot add backup cm", err)
+	}
```

Comment:
Why is that one function special and isn't part of the `backupResources` seems to be the same thing?